File: Advanced Programming Methods/Toy Language/src/Model/openRFileStatement.java
package Model;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;

//open the file fileName - if fileName already opened, throw exception -
//otherwise, add <unique FD,pair<fileName, bufferedReadder>> pair to fileTable to keep evidence of opened files
//also, to acces this file, add <varFileId, unique FD> pair to symbolTable
public class openRFileStatement implements IStatement {
    private String varFileId;
    private String fileName;
    private static int fileDescriptor = 1;

    public openRFileStatement(String varFileId, String fileName) {
        this.varFileId = varFileId;
        this.fileName = fileName;
    }

    @Override
    public ProgramState execute(ProgramState state) throws MyException, FileNotFoundException {
        IStack<IStatement> stk = state.getExeStack();
        IDictionary<Integer, Pair<String, BufferedReader>> fileTable = state.getFileTable();
        IDictionary<String, Integer> symTbl = state.getSymTable();

        BufferedReader myFile = null;

        for (Object el : fileTable.values()) {
            if (el instanceof Pair) {
                if (((Pair) el).getFirst() == fileName) {
                    throw new MyException("File Name already in the File Table");
                }
            }
        }

            myFile = new BufferedReader(new FileReader(fileName));
            fileTable.put(fileDescriptor, new Pair<>(fileName, myFile));
            symTbl.put(varFileId, fileDescriptor++);

            return state;
        }

    public String toString() {
        String res = "openRFile(";
        res = res.concat(varFileId);
        res = res.concat(",");
        res = res.concat(fileName);
        res = res.concat(")");
        return res;
    }
}

File: Advanced Programming Methods/Toy Language C#/A2/Model/Statements/openRFileStatement.cs
using System;
using System.Collections.Generic;
using A2.UI;
using A2.Model.Data_Structures;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace A2.Model.Statements
{
    //open the file fileName - if fileName already opened, throw exception -
    //otherwise, add <unique FD,pair<fileName, bufferedReadder>> pair to fileTable to keep evidence of opened files
    //also, to acces this file, add <varFileId, unique FD> pair to symbolTable
    public class openRFileStatement : IStatement {
        private String varFileId;
        private String fileName;
        private static int fileDescriptor = 1;

        public openRFileStatement(String varFileId, String fileName)
        {
            this.varFileId = varFileId;
            this.fileName = fileName;
        }
        public ProgramState execute(ProgramState state) {
            IStack<IStatement> stk = state.getExeStack();
            Data_Structures.IFileTable<int, KeyValuePair<string, StreamReader>> fileTable = state.getFileTable();
            Data_Structures.IDictionary<string, int> symTbl = state.getSymTable();

            StreamReader myFile = null;

            foreach (KeyValuePair<string, StreamReader> el in fileTable.values()) {
                    if (el.Key == fileName) {
                        throw new MyException("File Name already in the File Table");
                    }
            }

            myFile = new StreamReader(fileName);
            fileTable.put(fileDescriptor, new KeyValuePair<string, StreamReader>(fileName, myFile));
            symTbl.put(varFileId, fileDescriptor++);

            return state;
        }

            public override string ToString()
        {
            string res = "openRFile(";
            res += varFileId;
            res += ",";
            res += fileName;
            res += ")";
            return res;
        }
    }
}
